### PR TITLE
feat: add sync progress timeline

### PR DESCRIPTION
**Pull request by hiroppy**
# Summary

MoneyForward synchronization now exposes a durable, typed progress timeline from the crawler through the web dashboard.

- persists the latest running, successful, or failed run atomically in data/crawler-run-state.json
- records current work, waiting reasons, group and month metadata, warnings, and safe failure reasons
- proxies and validates the strict discriminated state at the web boundary
- shows a short desktop-only running or failed pill and keeps mobile icon-only
- opens one responsive timeline popup from running and failed refresh controls
- retries a failed synchronization from the popup
- adds unit, state, API, responsive Storybook, and runtime regression coverage

## Linear Issue

- [HIR-127](https://linear.app/hiroppy/issue/HIR-127/同期進捗タイムライン表示)

### Selected Quality Characteristics

| Quality characteristic | Why it is relevant | Acceptance criteria |
| ---------------------- | ------------------ | ------------------- |
| Functional suitability | The crawler, API, and UI must agree on every run state and step | current, timeline, waitingFor, progress, reason, retry, and terminal state behave end to end |
| Reliability | Polling can miss failures and concurrent starts can corrupt state | Atomic latest-state persistence, lock guards, stale-lock handling, and reload recovery are verified |
| Usability | Desktop and mobile have intentionally different header affordances | Desktop uses only a short pill; mobile remains icon-only; both open the same bounded popup |
| Interaction capability and accessibility | Status, progress, errors, and recovery must be perceivable and operable | Semantic progress, state-specific accessible names, contrast, popup bounds, and retry are verified |
| Security | Persisted errors must not expose raw credentials or malformed metadata | Runtime validation and safe reason normalization reject or redact unsafe input |

### Test Techniques

| Test technique | Selection rationale | Expected coverage |
| -------------- | ------------------- | ----------------- |
| Equivalence partitioning | Run behavior partitions into unavailable, idle, running, success, and failed | API and UI behavior for each meaningful state |
| Decision table testing | Click behavior depends on availability, active state, terminal state, and viewport | Start, open popup, retry, pill visibility, and disabled behavior |
| State transition testing | The feature is a synchronization lifecycle | start to running or waiting to success or failed to retry, including reload |
| Boundary value analysis | Counters, durations, viewport width, timestamps, and environment values have strict boundaries | Zero progress, finite durations, malformed timestamps, mobile width, and invalid wait configuration |
| Branch testing | Lock, persistence, warning, failure, and stale-state branches are high risk | Successful and failed phase boundaries, stale lock races, and refresh timeout continuation |
| Error guessing | Browser navigation, selectors, writes, cleanup, and callbacks can fail between normal checkpoints | Safe failure timeline items without orphan running steps or raw secret leakage |

### Primary Test Conditions

- A new run overwrites the previous failed state and begins at 0/10.
- Group discovery, refresh, global data, group data, month history, database, analytics, notification, and web refresh report user-facing steps.
- Refresh waiting retains remaining count and incomplete anonymous institution fixtures; timeout is a warning that allows the run to continue.
- Initial group or history navigation, later month scraping, cleanup, state writes, and detached runs leave a deterministic failed state.
- Page reload restores the latest running or failed timeline.
- Running and failed refresh clicks open the popup without a duplicate start; failed retry sends exactly one new request.
- Desktop 1280 by 900 shows the compact counter pill without header overflow.
- Mobile 390 by 844 hides the pill, keeps the icon entry, and contains the popup within the viewport.

### Out-of-Scope Risks

- Live MoneyForward scraping was not executed because it requires private credentials and mutates external account refresh state.
- Dashboard-wide authentication remains the existing private deployment trust boundary. Hardening all dashboard and crawler endpoints is tracked separately in HIR-132.
- Vercel preview deployment is rate limited by the free-tier daily deployment quota and is not a required code check.

## Verification

- [x] Unit tests
- [x] Type checking
- [x] Lint
- [x] Storybook tests
- [ ] Repository Web E2E suite
- [x] Manual runtime verification

### Commands Executed

- pnpm test — 6 of 6 tasks passed; crawler 224 tests and web unit plus Storybook 726 tests at the full-suite checkpoint
- pnpm --filter @mf-dashboard/crawler test — 26 files and 227 tests passed after final QA synchronization
- pnpm --filter @mf-dashboard/web test:unit — 24 files and 393 tests passed
- pnpm --filter @mf-dashboard/web test:storybook — 74 files and 333 tests passed in Chromium with accessibility checks
- pnpm turbo typecheck — 9 of 9 packages passed
- pnpm lint — passed
- pnpm format:check — 499 files passed
- pnpm knip — no unused exports introduced; existing lefthook unused devDependency remains
- Integrated Playwright runtime walkthrough — desktop running popup and mobile failed retry passed

### Runtime Media

Desktop running timeline popup:

Desktop running timeline

Mobile failed timeline popup and retry:

Mobile failed timeline

### Checks Not Run

- Repository Web E2E suite was not rerun. The changed responsive interaction path was exercised through Storybook Chromium tests and the integrated Playwright runtime walkthrough.
- Live crawler E2E was not run because it requires private MoneyForward credentials and performs external synchronization.

## Summary by CodeRabbit

* **New Features**
  * Added persisted crawler run progress with step timelines/current step.
  * Added month-level progress tracking for cash-flow history in history mode.
  * Added refresh UI improvements showing richer sync state and standardized retry behavior.
  * Added support for rendering refresh status with typed decoding and improved popover accessibility labeling.
* **Bug Fixes**
  * Improved error normalization and progress-aware handling for authentication, scraping phases, and notifications.
  * Strengthened run-lock reconciliation and terminal state persistence for consistent timelines.
* **Security**
  * Tightened same-origin validation for crawler refresh reads; standardized 503 unavailable responses.
* **Tests**
  * Expanded unit/e2e/UI test coverage across progress persistence, refresh/status handling, and edge cases.

### CODE REVIEW

**Review by hiroppy — COMMENTED** [review]
notification同じでpopupでよい

File: apps/crawler/src/crawler-run-lock.ts [hunk 1]
```diff
@@ -498,15 +566,18 @@ export async function getCrawlerRunState(
       };
 
   if (!isStaleSnapshot(snapshot, resolved)) {
+    if (record && progressState?.runId === record.id) {
+      return { ...progressState, running: true, pid: record.pid };
+    }
```
[the remark this answers is not in the document]
Comment (hiroppy) [reply apps/crawler/src/crawler-run-lock.ts:566]: 対応しました（5ee207c）。matching lock が残っていても persisted runStatus が success/failed なら terminal snapshot を running false / pid null で返し、running snapshot のみ lock flag を合成します。lock 解放前の success state を再現する回帰テストを追加し、crawler 228件、web unit 393件、typecheck 9/9、lint、formatを再検証しました。併せて review summary の型 re-export 指摘も解消し、ActionIcons は crawler-run-state から直接 type import します。

File: apps/crawler/src/server.ts [hunk 2]
```diff
@@ -27,25 +39,33 @@ function methodNotAllowed(response: ServerResponse): void {
...
     } finally {
       await lock.release();
     }
-  })();
+  })().catch((err) => {
+    error("Manual crawler run finalization failed:", err);
+  });
 
-  return state;
+  return { ...progress.getState(), running: true, pid: lock.record.pid };
```
Comment (hiroppy) [inline apps/crawler/src/server.ts:69]: `progress.getState()` now includes timeline metadata such as group names and `incompleteAccounts` account/service names, but the crawler trigger server is still unauthenticated and binds to `0.0.0.0`. In Compose it is only `expose`d internally, but local/non-Compose runs can leak these MoneyForward details to the LAN via `/runs` and `/status`. Please either bind this server to an internal/localhost interface, require a shared token, or redact the metadata from the direct crawler endpoint before returning it.

File: apps/crawler/src/crawler-run-lock.ts [hunk 3]
```diff
@@ -498,15 +566,21 @@ export async function getCrawlerRunState(
       };
 
   if (!isStaleSnapshot(snapshot, resolved)) {
+    if (record && progressState?.runId === record.id) {
+      if (progressState.runStatus !== "running") {
+        return { ...progressState, running: false, pid: null };
+      }
+      return { ...progressState, running: true, pid: record.pid };
```
Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts:565]: `getCrawlerRunState()` now returns the full persisted progress snapshot, including `timeline` metadata such as group names and refresh `incompleteAccounts`, and `server.ts` serves that object directly from unauthenticated `/status` (the crawler server still listens on `0.0.0.0`). That expands the crawler trigger API from a coarse lock status to potentially sensitive MoneyForward/account metadata. Please either keep the crawler server bound/authenticated for the web app only, or sanitize the direct crawler API response so external callers cannot retrieve timeline metadata.

File: apps/crawler/src/server.ts (hunk 2 above)
Comment (hiroppy) [reply apps/crawler/src/server.ts:69]: 対応しました（23bd49a）。crawler trigger server は通常起動時に 127.0.0.1 へ bind し、local/non-Compose 実行で timeline metadata を LAN に公開しないよう変更しました。Compose は web→crawler の internal service 通信を維持するため CRAWLER_HOST=0.0.0.0 を明示しています。default bind の回帰テストを追加し、crawler 230 tests、typecheck 9/9、compose config、lint、format を再検証しました。

File: apps/crawler/src/crawler-run-lock.ts (hunk 3 above)
Comment (hiroppy) [reply apps/crawler/src/crawler-run-lock.ts:565]: 対応しました（23bd49a）。status JSON を redaction せず web の group/month 表示契約を維持したまま、crawler server の default bind を 127.0.0.1 に制限しました。0.0.0.0 は外部 publish のない Compose internal network でのみ明示指定します。default bind regression test と crawler 全 230 tests、typecheck、compose config、lint、format は green です。

File: apps/crawler/src/crawler-run-lock.ts [hunk 4]
```diff
@@ -584,12 +658,24 @@ export async function acquireCrawlerRunLock(
...
+    await lock.release();
+    throw err;
+  }
   try {
-    return await task();
+    const result = await task(progress);
+    await progress.finish("success");
+    return result;
+  } catch (err) {
+    await progress.finish("failed");
```
Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts]: `task(progress)` の失敗後に `progress.finish("failed")` も失敗すると、ここで投げ直されるのは状態保存側のエラーになり、実際の crawler 失敗原因が失われます。manual run 側の `recordManualRunFailure()` と同様に、terminal state の保存失敗は secondary として扱い、元の `err` を保持/合成して返すようにしてください。

Comment (hiroppy) [reply apps/crawler/src/crawler-run-lock.ts]: 対応しました（cf1d6e1）。task の元エラーを `cause` と AggregateError.errors[0] に保持し、`progress.finish("failed")` の persistence error を errors[1] に合成して secondary failure として扱います。両方が失敗しても finally の lock release が完了し、次 run が acquire できる回帰テストを追加しました。crawler 231 tests、typecheck 9/9、lint、format は green です。

File: apps/crawler/src/crawler-run-lock.ts [hunk 5]
```diff
@@ -206,6 +262,9 @@ function isStaleLock(
     if (currentPidStartedAt && currentPidStartedAt !== record.pidStartedAt) {
       return true;
     }
+    if (currentPidStartedAt === record.pidStartedAt) {
```
Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts]: `startCrawlerRun()` now creates the lock with the long-lived crawler server PID, so this early `return false` means a manual run whose background task hangs (or never reaches `lock.release()`) will never become stale as long as the server process stays alive. Subsequent `/runs` calls will keep seeing the old lock as active indefinitely. Keep the PID-start check for PID reuse, but still fall through to the age/heartbeat-based stale check when the same process owns the lock.

Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts]: `startCrawlerRun()` now creates the lock with the long-lived crawler server PID, so returning `false` whenever that same PID is still alive means a manual run whose background task hangs or never reaches `lock.release()` will never become stale while the server process stays up. Subsequent `/runs` calls will keep seeing the old lock as active indefinitely. Keep the PID-start mismatch check for PID reuse, but still fall through to the age/heartbeat-based stale check when the same process owns the lock.

**Comment by hiroppy** [comment]
Commit: https://github.com/hiroppy/mf-dashboard/commit/9f262e0e849e64a1dfd79e86d0a844d2f312855e

### デスクトップで同期中の進捗ピルからタイムラインを確認できる

desktop-running-timeline

### モバイルでもアイコンから同期タイムラインを表示できる

mobile-running-timeline

### 失敗状態で理由と再試行ボタンを確認できる

mobile-failed-retry-timeline

**Comment by hiroppy** [comment]
Commit: https://github.com/hiroppy/mf-dashboard/commit/f6bccaffe03b18ac05ea12231331bbb827785091

### デスクトップで同期中の進捗ピルからタイムラインを確認できる

desktop-running-timeline

### モバイルでもアイコンから同期タイムラインを表示できる

mobile-running-timeline

### 失敗状態で理由と再試行ボタンを確認できる

mobile-failed-retry-timeline

File: apps/web/src/components/layout/action-icons.tsx [hunk 6]
```diff
@@ -79,14 +75,18 @@ function RefreshButton({ iconSize }: { iconSize: string }) {
         const nextStatus = await readCrawlerRefreshStatus();
         if (isMounted) {
           setState({ ...nextStatus, isPending: false });
```
Comment (hiroppy) [inline apps/web/src/components/layout/action-icons.tsx:84]: `updateStatus()` and `startRefresh()` can both have in-flight requests, but their responses are applied unconditionally. If a polling GET was started before the user clicks refresh and resolves after the optimistic `running: true` update, this line can overwrite the UI back to idle and reset `wasRunningRef`; conversely a late POST response can force `running: true` after a newer poll has already observed completion. Please serialize/cancel stale requests (or use a monotonically increasing request generation/run id and ignore older responses) so the timeline state only moves forward for the latest request.

File: apps/crawler/src/crawler-run-lock.ts [hunk 7]
```diff
@@ -498,15 +564,21 @@ export async function getCrawlerRunState(
       };
 
   if (!isStaleSnapshot(snapshot, resolved)) {
+    if (record && progressState?.runId === record.id) {
+      if (progressState.runStatus !== "running") {
```
Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts]: When the progress file has reached `success`/`failed` but the lock file still exists, `getCrawlerRunState()` reports `running: false`. A concurrent `/runs` call then hits `lockExists`, sees `!state.running`, and recursively retries `acquireCrawlerRunLock()` while the same lock is still present; if `lock.release()` is delayed or fails, this can spin/stack-recurse instead of returning a normal 409/running state. Please treat the lock as authoritative until it is released (or replace the recursive retry with a bounded wait/backoff that only proceeds after the lock is actually gone).

File: apps/crawler/src/crawler-run-lock.ts [hunk 8]
```diff
@@ -498,15 +561,18 @@ export async function getCrawlerRunState(
       };
 
   if (!isStaleSnapshot(snapshot, resolved)) {
+    if (record && progressState?.runId === record.id) {
+      return { ...progressState, running: true, pid: record.pid };
+    }
     return snapshotState;
   }
```
Comment (hiroppy) [inline apps/crawler/src/crawler-run-lock.ts:568]: `getCrawlerRunState()` reports `running: false` whenever the persisted progress for the live lock is terminal, but the lock file can still be present if `lock.release()` fails after `progress.finish("success")` in the detached manual run. In that state, the next `acquireCrawlerRunLock()` hits `EEXIST`, calls `getCrawlerRunState()`, sees `running: false`, and recursively retries forever while the still-live PID prevents the lock from becoming stale. Please treat a valid live lock as running regardless of terminal progress state, or remove/bound the retry path, and add a regression for terminal progress + unreleased lock.
